File: app/src/main/java/com/example/merchio/DetailProductActivity.java
package com.example.merchio;

import android.content.Intent;
import android.os.Bundle;
import android.widget.Button;
import android.widget.ImageButton;
import android.widget.ImageView;
import android.widget.TextView;
import android.widget.Toast;

import androidx.appcompat.app.AppCompatActivity;

import com.bumptech.glide.Glide;
import com.example.merchio.fragments.ProductActionBottomSheetDialogFragment;
import com.example.merchio.models.Product;

import java.text.NumberFormat;
import java.util.Locale;

public class DetailProductActivity extends AppCompatActivity
        implements ProductActionBottomSheetDialogFragment.Listener {

    private Product product;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_detail_product);

        product = getIntent().getParcelableExtra("product");

        if (product == null) {
            finish();
            return;
        }

        ImageView imgProduct = findViewById(R.id.imgProduct);
        TextView tvName = findViewById(R.id.tvName);
        TextView tvPrice = findViewById(R.id.tvPrice);
        TextView tvType = findViewById(R.id.tvType);
        TextView tvDescription = findViewById(R.id.tvDescription);

        ImageButton btnCart = findViewById(R.id.btnCart);
        Button btnBuyNow = findViewById(R.id.btnBuyNow);
        TextView btnBack = findViewById(R.id.btn_back);

        btnBack.setOnClickListener(v -> finish());

        tvName.setText(product.getName());
        tvPrice.setText(formatRupiah(product.getPrice()));
        tvType.setText("Type: " + (product.getType() != null ? product.getType() : "-"));
        tvDescription.setText(product.getDescription());

        Glide.with(this)
                .load(product.getImageUrl())
                .into(imgProduct);

        btnCart.setOnClickListener(v -> {
            ProductActionBottomSheetDialogFragment
                    .newInstance(product, ProductActionBottomSheetDialogFragment.MODE_ADD_TO_CART)
                    .show(getSupportFragmentManager(), "product_action_add");
        });

        btnBuyNow.setOnClickListener(v -> {
            ProductActionBottomSheetDialogFragment
                    .newInstance(product, ProductActionBottomSheetDialogFragment.MODE_BUY_NOW)
                    .show(getSupportFragmentManager(), "product_action_buy");
        });
    }

    @Override
    public void onBuyNowRequested(Product product,
                                  String selectedType,
                                  int quantity) {

        Intent intent =
                new Intent(
                        this,
                        CheckoutActivity.class
                );

        intent.putExtra(
                "buy_now",
                true
        );

        intent.putExtra(
                "product",
                product
        );

        intent.putExtra(
                "qty",
                quantity
        );

        intent.putExtra(
                "type",
                selectedType
        );

        startActivity(intent);
    }

    private String formatRupiah(int price) {
        NumberFormat formatter = NumberFormat.getCurrencyInstance(new Locale("id", "ID"));
        return formatter.format(price).replace(",00", "");
    }
}

File: app/src/main/java/com/example/merchio/DetailProductActivity.kt
package com.example.merchio

import android.content.Intent
import android.os.Bundle
import android.view.View
import android.widget.ImageView
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.lifecycleScope
import com.example.merchio.data.model.ProductDto
import com.example.merchio.data.repository.CartRepository
import com.example.merchio.data.repository.ProductRepository
import com.example.merchio.utils.CurrencyFormatter
import com.example.merchio.utils.IntentKeys
import com.example.merchio.utils.loadMerchioImage
import com.example.merchio.utils.toast
import kotlinx.coroutines.launch

class DetailProductActivity : AppCompatActivity() {
    private val productRepository = ProductRepository()
    private val cartRepository = CartRepository()
    private var product: ProductDto? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_detail_product)
        findViewById<View>(R.id.btn_back).setOnClickListener { finish() }
        findViewById<View>(R.id.btnCart).setOnClickListener { addToCart() }
        findViewById<View>(R.id.btnBuyNow).setOnClickListener { buyNow() }
        loadProduct()
    }

    override fun onResume() { super.onResume(); if (product != null) loadProduct() }

    private fun loadProduct() {
        val id = intent.getStringExtra(IntentKeys.PRODUCT_ID) ?: return finish()
        lifecycleScope.launch {
            runCatching { productRepository.productById(id) }
                .onSuccess { data ->
                    product = data
                    findViewById<ImageView>(R.id.imgProduct).loadMerchioImage(data.imageUrl)
                    findViewById<TextView>(R.id.tvName).text = data.name
                    findViewById<TextView>(R.id.tvPrice).text = CurrencyFormatter.rupiah(data.price)
                    findViewById<TextView>(R.id.tvType).text = "Stock ${data.stock} • Sold ${data.soldCount}"
                    findViewById<TextView>(R.id.tvDescription).text = data.description ?: "No description"
                }
                .onFailure { toast(it.message ?: "Produk tidak ditemukan"); finish() }
        }
    }

    private fun addToCart() {
        val item = product ?: return
        if (item.stock <= 0) { toast("Stock habis"); return }
        lifecycleScope.launch {
            runCatching { cartRepository.addToCart(item.id) }
                .onSuccess { toast("Ditambahkan ke cart") }
                .onFailure { toast(it.message ?: "Gagal tambah cart") }
        }
    }

    private fun buyNow() {
        val item = product ?: return
        if (item.stock <= 0) { toast("Stock habis"); return }
        startActivity(Intent(this, CheckoutActivity::class.java)
            .putExtra(IntentKeys.CHECKOUT_MODE, IntentKeys.MODE_BUY_NOW)
            .putExtra(IntentKeys.PRODUCT_ID, item.id)
            .putExtra(IntentKeys.QUANTITY, 1))
    }
}
